### PR TITLE
Adds method to sample active staking providers using TACo child app

### PR DESCRIPTION
**Pull request by vzotova**
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
Adds `getActiveStakingProviders` to `TACoChildApplication`

**Why it's needed:**
This method works using operator confirmations, and confirmations use bridge from polygon to mainnet. This direction needs bot or changes in dashboard. To decrease priority for that direction we can use child app for sampling

### CODE REVIEW

File: contracts/contracts/coordination/TACoChildApplication.sol [hunk 1]
```diff
@@ -123,6 +130,56 @@ contract TACoChildApplication is ITACoRootToChild, ITACoChildApplication, Initia
...
+     * @notice Get the value of authorized tokens for active providers as well as providers and their authorized tokens
+     * @param _startIndex Start index for looking in providers array
+     * @param _maxStakingProviders Max providers for looking, if set 0 then all will be used
+     * @return allAuthorizedTokens Sum of authorized tokens for active providers
+     * @return activeStakingProviders Array of providers and their authorized tokens.
+     * Providers addresses stored as uint256
+     * @dev Note that activeStakingProviders[0] is an array of uint256, but you want addresses.
+     * Careful when used directly!
+     */
+    function getActiveStakingProviders(
```
Comment (derekpierre) [inline contracts/contracts/coordination/TACoChildApplication.sol:151]: In `nucypher-porter` and `nucypher` we currently do sampling using the same function on `TACoApplication` contract instead of `TACoChildApplication`. With this change, instead, we can sample using `TACoChildApplication`? Is there any difference between sampling using root vs child, other than using different chains? Is one potentially more up-to-date than the other?

I'm asking because, at the moment, Porter doesn't use/need any polygon endpoint, it only needs an eth endpoint. If using `TACoChildApplication` is better for some reason, then Porter will need to be updated to use polygon.

Comment (vzotova) [reply contracts/contracts/coordination/TACoChildApplication.sol:151]: yep, that's alternative to method in TACo root app until bot to sync from polygon to mainnet will be ready. 
Thing is - sampling returns only confirmed providers/operators. And without bot - all providers in root app will be unconfirmed. And considering that child has all data already - that's easiest option to start with

File: contracts/contracts/coordination/TACoChildApplication.sol [hunk 2]
```diff
@@ -123,6 +130,56 @@ contract TACoChildApplication is ITACoRootToChild, ITACoChildApplication, Initia
...
+
+        uint256 resultIndex = 0;
+        for (uint256 i = _startIndex; i < endIndex; i++) {
+            address stakingProvider = stakingProviders[i];
+            StakingProviderInfo storage info = stakingProviderInfo[stakingProvider];
+            uint256 eligibleAmount = info.authorized;
+            if (eligibleAmount < minimumAuthorization || !info.operatorConfirmed) {
+                continue;
+            }
+            activeStakingProviders[resultIndex][0] = uint256(uint160(stakingProvider));
```
Comment (manumonti) [inline contracts/contracts/coordination/TACoChildApplication.sol:175]: Just out of curiosity: why returning `uint256` instead of `address`?

**Review by manumonti — APPROVED** [review]
LGTM ✌️! I've just left a question.

File: contracts/contracts/coordination/TACoChildApplication.sol (hunk 2 above)
Comment (cygnusv) [reply contracts/contracts/coordination/TACoChildApplication.sol:175]: Because it's an array with 2 columns, one for address and the other for amount, so `uint256` seems like the more "common" type. 

@vzotova knowing that the address plus the amount (uint96) fit exactly in 1 word, how about returning a single uint256 value? This has the advantage that it would reduce the likelihood of using pagination by 50%, at the expense of requiring a very minor processing on the client side.

File: contracts/contracts/coordination/TACoChildApplication.sol [hunk 3]
```diff
@@ -123,6 +130,56 @@ contract TACoChildApplication is ITACoRootToChild, ITACoChildApplication, Initia
...
+            activeStakingProviders[resultIndex++][1] = eligibleAmount;
+            allAuthorizedTokens += eligibleAmount;
+        }
+        assembly {
+            mstore(activeStakingProviders, resultIndex)
```
Comment (cygnusv) [inline contracts/contracts/coordination/TACoChildApplication.sol:180]: Why is this necessary?

Comment (vzotova) [reply contracts/contracts/coordination/TACoChildApplication.sol:180]: to cut empty spaces in returning array, otherwise it will always be `endIndex - _startIndex` with zeros in the end

File: contracts/contracts/coordination/TACoChildApplication.sol [hunk 4]
```diff
@@ -19,6 +19,7 @@ contract TACoChildApplication is ITACoRootToChild, ITACoChildApplication, Initia
         bool operatorConfirmed;
         uint96 authorized;
         // TODO: what about undelegations etc?
+        uint256 index; // index in stakingProviders array + 1
```
[the remark this answers is not in the document]
Comment (vzotova) [reply contracts/contracts/coordination/TACoChildApplication.sol]: I tried 2 slots and it gives almost no gas savings (< 100 gas in average tx). Because it mostly about how and when we access those slots and less slots not always better. But I don't mind to make those changes
✔️

File: contracts/contracts/coordination/TACoChildApplication.sol (hunk 2 above)
Comment (vzotova) [reply contracts/contracts/coordination/TACoChildApplication.sol]: ✔️

**Review by cygnusv — APPROVED** [review]
